vine/src/features/type_.rs

```diff
@@ -67,14 +67,32 @@ impl<'core> Charter<'core, '_> {
 
 impl<'core> Resolver<'core, '_> {
   pub(crate) fn resolve_type_alias(&mut self, alias_id: TypeAliasId) {
-    if let Some(Some(_)) = self.sigs.type_aliases.get(alias_id) {
-      return;
+    match self.sigs.type_aliases.get(alias_id) {
+      Some(TypeAliasState::Resolved(_)) => return,
+      Some(TypeAliasState::Resolving) => {
+        // Cycle detected
+        let alias_def = &self.chart.type_aliases[alias_id];
+        self.core.report(Diag::RecursiveTypeAlias { span: alias_def.ty.span });
+        let slot = self.sigs.type_aliases.get_or_extend(alias_id);
+        let error_type =
+          self.types.error(self.core.report(Diag::RecursiveTypeAlias { span: alias_def.ty.span }));
+        *slot = TypeAliasState::Resolved(
+          self.types.export(|t| TypeAliasSig { ty: t.transfer(&error_type) }),
+        );
+        return;
+      }
+      _ => {}
```

Comment:
I think it would be cleaner to move the remaining logic of the function into this arm and remove the two early `return`s.